### PR TITLE
Edge detection guide

### PR DESCRIPTION
**Pull request by kapil-varshney**
This adds the Deep learning - Edge Detection guide.

### CODE REVIEW

**Comment by kapil-varshney** [comment]
@vikashranjan26 Can you please review the technicals details in the guide. Please let me know if there's anything that's not mentioned correctly or that I missed out on and should be included.

File: guide/14-deep-learning/edge-detection.ipynb [hunk 1]
```diff
@@ -0,0 +1,159 @@
+{
```
Comment (vikashranjan26) [inline guide/14-deep-learning/edge-detection.ipynb:1]: <p>In our model, I have supported all vgg and resnet backbone so please mention in the guide as well.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/895/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

**Comment by priyankatuteja** [comment]
@BP-Ent Could you please help with the review.

**Comment by priyankatuteja** [comment]
@kapil-varshney LGTM. You can incorporate changes that @vikashranjan26 mentioned.

**Review by vikashranjan26 — APPROVED** [review]
LGTM

File: guide/14-deep-learning/edge-detection.ipynb [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
```
Comment (BP-Ent) [inline guide/14-deep-learning/edge-detection.ipynb:1]: <p>Edge Detection is the identification of edges and object boundaries in an image. As a human, we can easily identify the edges and object boundaries in images due to our highly evolved visual perception and prior context and understanding of the world.  A computer algorithm, in contrast, would identify these edges by identifying changes in contrasts, gradients, colors, etc. within an image, which are represented by different pixel values.  Edge Detection has multiple use cases across many domains; for example, Edge Detection can be used to identify land parcel boundaries in satellite imagery.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/895/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (BP-Ent) [inline guide/14-deep-learning/edge-detection.ipynb:1]: <p>Edge Detection is a classical computer vision problem. There have been many algorithms in the past that have worked well, to a certain degree, for edge detection. Most of these employed well-researched filters or operators that worked in most cases. The Canny Edge Detection [2] technique has been one of the most popular ones. This technique involves multiple stages, including Gaussian blurring, gradient filtering, non-maxima suppression, etc. Many other techniques use hand-crafted features to detect the edges. Later, algorithms such as Structured Forest for Fast Edge Detection, which use machine learning with hand-crafted features, grew in popularity as they were more accurate and relatively faster than their predecessors.[3]  </p><p>In the last few years, with advancements in deep learning, many CNN-based models have been developed to solve edge detection problems. When using these deep learning models, Edge Detection can be considered a special case of pixel classification (or image segmentation, as called in AI jargon).  In <code>arcgis.learn</code>, we have integrated two deep learning models for Edge Detection - HED and BDCN.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/895/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (BP-Ent) [inline guide/14-deep-learning/edge-detection.ipynb:1]: <p>HED is one of the earlier CNN-based models for edge detection. The model has two salient features that give the model its name, according to the authors. First, the model is 'holistic', as it takes an image as input and outputs another image (edge map). The model doesn't require hand-crafted features as inputs, rather, due to its architecture, it creates these features internally in the hidden layers. This property is inherited from the fully-convolutional networks. Second, the model is 'nested' and learns at multiple scales using deep supervised learning. This is achieved by taking side outputs at varying depths. [4]</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/895/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (BP-Ent) [inline guide/14-deep-learning/edge-detection.ipynb:1]: <p>The model uses VGGNet architecture as its base that is trimmed from its last pooling layer. The five convolution blocks have a side output, each of which helps in learning features at multiple scales. These side outputs are combined together to form a fusion layer, the weights of which are also learnable. The resulting fusion layer provides a unified output. This architecture of HED can be seen in Figure 2. Total loss is the sum of loss calculated at both the fusion layer and the side outputs. The loss function used in the model is the cross-entropy loss with the addition of a class-balancing weight for the side outputs. The class-balancing weight is introduced because the majority of the pixels in an edge map will not represent edges.[4]</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/895/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (BP-Ent) [inline guide/14-deep-learning/edge-detection.ipynb:1]: <p>Like HED, BDCN model uses a VGGNet as the base architecture. Each of the five convolutional blocks leads to a pooling layer and then evolves into an Incremental Detection Block (ID Block). These ID Blocks are comprised of several convolutional layers, followed by a Scale Enhancement Module (SEM). SEM is made of multiple parallel convolution layers using dilation at different rates to create receptive fields at varying scales. Using dilation is an effective way of creating varying receptive fields without increasing the total parameters. ID Blocks eventually provide two edge predictions, with one propagated from the shallow layer to the deep layer and the other from the deep layer to the shallow layer, thus forming a bi-directional cascade structure. The final prediction is a fusion of the intermediate edge map predictions. [1]</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/895/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (BP-Ent) [inline guide/14-deep-learning/edge-detection.ipynb:1]: <p>You can optionally provide a <code>backbone</code> parameter, the default value for which is <code>vgg19</code>. In our implementation of Edge Detection models, ResNets can also be used apart from VGGNets as the backbones. </p><p> </p><p><strong>Q: Should this be "<em>ResNets can also be used as the backbone, rather than VGGNets</em>"?</strong></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/895/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

**Review by BP-Ent — CHANGES_REQUESTED** [review]
Suggested changes made on ReviewNB

File: guide/14-deep-learning/edge-detection.ipynb (hunk 2 above)
Comment (kapil-varshney) [reply guide/14-deep-learning/edge-detection.ipynb:1]: <p>I mean to say that both ResNets and VGGNets can be used. The authors made it work with VGG, but additionally, we have provided the option of ResNets also.</p>

Comment (kapil-varshney) [reply guide/14-deep-learning/edge-detection.ipynb:1]: <p>"You can optionally provide a <code>backbone</code> parameter, the default value for which is <code>vgg19</code>. In our implementation of Edge Detection models, <strong>ResNets can also be used as the backbone, in addition to VGGNets.</strong>"</p>

**Review by priyankatuteja — APPROVED** [review]

**Review by priyankatuteja — CHANGES_REQUESTED** [review]
@kapil-varshney Should we rename the notebook to 'Edge Detection with arcgis.learn'

**Comment by kapil-varshney** [comment]
> @kapil-varshney Should we rename the notebook to 'Edge Detection with arcgis.learn'

Yes, that sounds better. I have updated the title of the notebook.

**Review by priyankatuteja — APPROVED** [review]

File: guide/14-deep-learning/edge-detection.ipynb (hunk 2 above)
Comment (AtmaMani) [inline guide/14-deep-learning/edge-detection.ipynb:1]: <p>Replace title with the title of the guide. We need this for better search engine indexing and for browsers to title the tabs</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/895/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

**Review by AtmaMani — CHANGES_REQUESTED** [review]
Title needs to be changed, looks good otherwise

File: guide/14-deep-learning/edge-detection.ipynb (hunk 2 above)
Comment (kapil-varshney) [reply guide/14-deep-learning/edge-detection.ipynb:1]: I have added a title to the guide and renamed the notebook with the same name. Is that what you meant?

**Review by AtmaMani — APPROVED** [review]
